## Day 14. Task 01
## Регуляризация
### 0. Импорты

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
import warnings
warnings.filterwarnings('ignore')

### 1. Препроцессинг
 
1. Загрузите в датафрейм файл dayofweek.csv, которым вы пользовались в одном из предыдущих проектов.
2. Воспользуйтесь train_test_split с параметрами test_size=0.2, random_state=21. Получите X_train, y_train, X_test, y_test. Добавьте еще один параметр stratify.

In [2]:
df = pd.read_csv('dayofweek.csv')

In [3]:
df

numTrials      hour  dayofweek  user_0  user_1  user_10  user_11  \
0     -0.788667 -2.562352          4       0       0        0        0   
1     -0.756764 -2.562352          4       0       0        0        0   
2     -0.724861 -2.562352          4       0       0        0        0   
3     -0.692958 -2.562352          4       0       0        0        0   
4     -0.661055 -2.562352          4       0       0        0        0   
...         ...       ...        ...     ...     ...      ...      ...   
1681  -0.533442  0.945382          3       0       0        0        0   
1682  -0.629151  0.945382          3       0       1        0        0   
1683  -0.597248  0.945382          3       0       1        0        0   
1684  -0.565345  0.945382          3       0       1        0        0   
1685  -0.533442  0.945382          3       0       1        0        0   

      user_12  user_13  user_14  ...  lab02  lab03  lab03s  lab05s  laba04  \
0           0        0        0  ...      0      0       0       0       0   
1           0        0        0  ...      0      0       0       0       0   
2           0        0        0  ...      0      0       0       0       0   
3           0        0        0  ...      0      0       0       0       0   
4           0        0        0  ...      0      0       0       0       0   
...       ...      ...      ...  ...    ...    ...     ...     ...     ...   
1681        0        0        0  ...      0      0       0       0       0   
1682        0        0        0  ...      0      0       0       0       0   
1683        0        0        0  ...      0      0       0       0       0   
1684        0        0        0  ...      0      0       0       0       0   
1685        0        0        0  ...      0      0       0       0       0   

      laba04s  laba05  laba06  laba06s  project1  
0           0       0       0        0         1  
1           0       0       0        0         1  
2           0       0       0        0         1  
3           0       0       0        0         1  
4           0       0       0        0         1  
...       ...     ...     ...      ...       ...  
1681        0       0       0        1         0  
1682        0       0       0        1         0  
1683        0       0       0        1         0  
1684        0       0       0        1         0  
1685        0       0       0        1         0  

[1686 rows x 44 columns]

In [5]:
X = df.drop(columns=['dayofweek'])
X

numTrials      hour  user_0  user_1  user_10  user_11  user_12  user_13  \
0     -0.788667 -2.562352       0       0        0        0        0        0   
1     -0.756764 -2.562352       0       0        0        0        0        0   
2     -0.724861 -2.562352       0       0        0        0        0        0   
3     -0.692958 -2.562352       0       0        0        0        0        0   
4     -0.661055 -2.562352       0       0        0        0        0        0   
...         ...       ...     ...     ...      ...      ...      ...      ...   
1681  -0.533442  0.945382       0       0        0        0        0        0   
1682  -0.629151  0.945382       0       1        0        0        0        0   
1683  -0.597248  0.945382       0       1        0        0        0        0   
1684  -0.565345  0.945382       0       1        0        0        0        0   
1685  -0.533442  0.945382       0       1        0        0        0        0   

      user_14  user_15  ...  lab02  lab03  lab03s  lab05s  laba04  laba04s  \
0           0        0  ...      0      0       0       0       0        0   
1           0        0  ...      0      0       0       0       0        0   
2           0        0  ...      0      0       0       0       0        0   
3           0        0  ...      0      0       0       0       0        0   
4           0        0  ...      0      0       0       0       0        0   
...       ...      ...  ...    ...    ...     ...     ...     ...      ...   
1681        0        0  ...      0      0       0       0       0        0   
1682        0        0  ...      0      0       0       0       0        0   
1683        0        0  ...      0      0       0       0       0        0   
1684        0        0  ...      0      0       0       0       0        0   
1685        0        0  ...      0      0       0       0       0        0   

      laba05  laba06  laba06s  project1  
0          0       0        0         1  
1          0       0        0         1  
2          0       0        0         1  
3          0       0        0         1  
4          0       0        0         1  
...      ...     ...      ...       ...  
1681       0       0        1         0  
1682       0       0        1         0  
1683       0       0        1         0  
1684       0       0        1         0  
1685       0       0        1         0  

[1686 rows x 43 columns]

In [6]:
y = df['dayofweek']
y

0       4
1       4
2       4
3       4
4       4
       ..
1681    3
1682    3
1683    3
1684    3
1685    3
Name: dayofweek, Length: 1686, dtype: int64

In [7]:
X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, test_size=0.2, random_state=21)

In [8]:
X_train

numTrials      hour  user_0  user_1  user_10  user_11  user_12  user_13  \
860   -0.724861 -0.691561       0       0        0        0        0        0   
385   -0.629151 -1.159259       0       0        0        0        0        0   
422   -0.788667 -1.393108       0       0        0        0        0        0   
326   -0.533442  0.945382       0       0        0        0        0        0   
714    0.678880 -0.925410       0       0        0        0        0        0   
...         ...       ...     ...     ...      ...      ...      ...      ...   
1413  -0.501539  0.009986       0       0        0        0        0        0   
1637  -0.756764 -0.457712       0       0        0        0        0        0   
1334   1.444556 -0.691561       0       0        0        0        0        0   
979    0.742686  0.711533       0       0        0        0        0        0   
1241   0.104622  0.945382       0       0        0        0        0        0   

      user_14  user_15  ...  lab02  lab03  lab03s  lab05s  laba04  laba04s  \
860         0        0  ...      0      0       0       0       0        0   
385         0        0  ...      0      0       0       0       0        0   
422         0        0  ...      0      0       0       1       0        0   
326         0        0  ...      0      0       0       0       1        0   
714         0        0  ...      0      0       0       0       0        0   
...       ...      ...  ...    ...    ...     ...     ...     ...      ...   
1413        0        0  ...      0      0       0       0       0        0   
1637        0        0  ...      0      0       0       0       0        0   
1334        0        0  ...      0      0       0       0       0        0   
979         0        0  ...      0      0       0       0       0        0   
1241        0        0  ...      0      0       0       0       0        0   

      laba05  laba06  laba06s  project1  
860        0       0        0         1  
385        1       0        0         0  
422        0       0        0         0  
326        0       0        0         0  
714        0       0        0         1  
...      ...     ...      ...       ...  
1413       0       0        0         1  
1637       0       0        1         0  
1334       0       0        0         1  
979        0       0        0         1  
1241       0       0        0         1  

[1348 rows x 43 columns]

In [9]:
y_train

860     6
385     4
422     5
326     6
714     1
       ..
1413    3
1637    1
1334    3
979     0
1241    2
Name: dayofweek, Length: 1348, dtype: int64

In [10]:
X_test

numTrials      hour  user_0  user_1  user_10  user_11  user_12  user_13  \
1087   1.316943  0.243835       0       0        0        0        0        0   
16    -0.788667 -0.691561       0       0        0        0        0        0   
563   -0.373926 -1.393108       0       0        0        0        1        0   
1381  -0.182507 -0.223863       0       0        0        0        0        0   
1199  -0.533442 -0.691561       0       0        0        0        0        0   
...         ...       ...     ...     ...      ...      ...      ...      ...   
1411   4.156328  0.009986       0       0        0        0        0        0   
1079   1.061718  0.243835       0       0        0        0        0        0   
1222  -0.118700  0.243835       0       0        0        0        0        0   
1064   0.774589  0.009986       0       0        0        0        0        0   
1243   0.359848  0.945382       0       0        0        0        0        0   

      user_14  user_15  ...  lab02  lab03  lab03s  lab05s  laba04  laba04s  \
1087        1        0  ...      0      0       0       0       0        0   
16          0        0  ...      0      0       0       0       0        1   
563         0        0  ...      0      0       0       0       0        0   
1381        0        0  ...      0      0       0       0       0        0   
1199        0        0  ...      0      0       0       0       0        0   
...       ...      ...  ...    ...    ...     ...     ...     ...      ...   
1411        0        0  ...      0      0       0       0       0        0   
1079        1        0  ...      0      0       0       0       0        0   
1222        0        0  ...      0      0       0       0       0        0   
1064        1        0  ...      0      0       0       0       0        0   
1243        0        0  ...      0      0       0       0       0        0   

      laba05  laba06  laba06s  project1  
1087       0       0        0         1  
16         0       0        0         0  
563        1       0        0         0  
1381       0       0        0         1  
1199       0       0        0         1  
...      ...     ...      ...       ...  
1411       0       0        0         1  
1079       0       0        0         1  
1222       0       0        0         1  
1064       0       0        0         1  
1243       0       0        0         1  

[338 rows x 43 columns]

In [11]:
y_test

1087    1
16      5
563     6
1381    3
1199    2
       ..
1411    3
1079    1
1222    2
1064    1
1243    2
Name: dayofweek, Length: 338, dtype: int64

In [12]:
y_train.value_counts(normalize=True)

3    0.234421
6    0.211424
1    0.162463
5    0.160979
2    0.088279
0    0.080861
4    0.061573
Name: dayofweek, dtype: float64

In [13]:
y_test.value_counts(normalize=True)

3    0.236686
6    0.210059
1    0.162722
5    0.159763
2    0.088757
0    0.079882
4    0.062130
Name: dayofweek, dtype: float64

### 2. Регуляризация логрегрессии
### a. Логрегрессия по умолчанию
 
1. Обучите baseline-модель, только используя параметры random_state=21, fit_intercept=False.
2. Воспользуетесь K-fold кросс-валидацией со стратификацией, количество сплитов для оценки accuracy = 10.
Результат кода, в котором вы обучили модель и оценили ее качество, должен быть точно такой же, как и ниже:  
train -  0.62902   |   valid -  0.59259  
train -  0.63479   |   valid -  0.56296  
train -  0.65622   |   valid -  0.61481  
train -  0.63397   |   valid -  0.57778  
train -  0.64056   |   valid -  0.59259  
train -  0.64138   |   valid -  0.65926  
train -  0.65952   |   valid -  0.56296  
train -  0.64333   |   valid -  0.59701  
train -  0.63674   |   valid -  0.62687  
Average accuracy on crossval is 0.60165  
Std is 0.02943  
Воспользуйтесь %%time, чтобы понять, насколько долго выполнялась эта ячейка с кодом.

In [16]:
def cross_val(estimator, X, y, n_splits):
    score = []
    skf = StratifiedKFold(n_splits=n_splits)
    for train_index, valid_index in skf.split(X, y):
        X_train, X_valid = X.iloc[train_index], X.iloc[valid_index]
        y_train, y_valid = y.iloc[train_index], y.iloc[valid_index]
        estimator.fit(X_train, y_train)
        score.append(accuracy_score(y_valid, estimator.predict(X_valid)))
        print(f'train  -  {accuracy_score(y_train, estimator.predict(X_train)):.5f}\t|\tvalid  -  {accuracy_score(y_valid, estimator.predict(X_valid)):.5f}')

    print(f'Average accuracy on crossval is {np.mean(score):.5f}')
    print(f'Std is {np.std(score):.5f}')

In [18]:
%%time
cross_val(LogisticRegression(random_state=21, fit_intercept=False), X_train, y_train, 10)

train  -  0.62902	|	valid  -  0.59259
train  -  0.64633	|	valid  -  0.62963
train  -  0.63479	|	valid  -  0.56296
train  -  0.65622	|	valid  -  0.61481
train  -  0.63397	|	valid  -  0.57778
train  -  0.64056	|	valid  -  0.59259
train  -  0.64138	|	valid  -  0.65926
train  -  0.65952	|	valid  -  0.56296
train  -  0.64333	|	valid  -  0.59701
train  -  0.63674	|	valid  -  0.62687
Average accuracy on crossval is 0.60165
Std is 0.02943
CPU times: total: 969 ms
Wall time: 277 ms


### b. Оптимизация параметров регуляризации
 В ячейках ниже попробуйте разные значения penalty: none, l1, l2. Если потребуется, то вы также можете менять значения solver. Посмотрите, как меняется качество модели.

In [22]:
solver_penalty = {'newton-cg': ['l2', 'none'],
                'lbfgs': ['l2', 'none'],
                'liblinear': ['l1', 'l2'],
                'sag': ['l2', 'none'],
                'saga': ['l1', 'l2', 'none']}

In [23]:
for solver in solver_penalty:
    for penalty in solver_penalty[solver]:
        print(f'solver = {solver}, penalty = {penalty}')
        cross_val(LogisticRegression(random_state=21, fit_intercept=False, penalty=penalty, solver=solver), X_train, y_train, 10)
        print()

solver = newton-cg, penalty = l2
train  -  0.62902	|	valid  -  0.59259
train  -  0.64633	|	valid  -  0.62963
train  -  0.63479	|	valid  -  0.56296
train  -  0.65622	|	valid  -  0.61481
train  -  0.63397	|	valid  -  0.57778
train  -  0.64056	|	valid  -  0.59259
train  -  0.64221	|	valid  -  0.65926
train  -  0.65952	|	valid  -  0.56296
train  -  0.64333	|	valid  -  0.59701
train  -  0.63674	|	valid  -  0.62687
Average accuracy on crossval is 0.60165
Std is 0.02943

solver = newton-cg, penalty = none
train  -  0.66694	|	valid  -  0.63704
train  -  0.65787	|	valid  -  0.65926
train  -  0.66694	|	valid  -  0.57778
train  -  0.66529	|	valid  -  0.62963
train  -  0.66694	|	valid  -  0.62222
train  -  0.65952	|	valid  -  0.57778
train  -  0.65045	|	valid  -  0.69630
train  -  0.68425	|	valid  -  0.61481
train  -  0.66474	|	valid  -  0.62687
train  -  0.65651	|	valid  -  0.60448
Average accuracy on crossval is 0.62462
Std is 0.03379

solver = lbfgs, penalty = l2
train  -  0.62902	|	valid  -  0

Best LogisticRegression: solver = lbfgs, penalty = none  
Average accuracy on crossval is 0.62537  
Std is 0.03302

### 3. Регуляризация SVM
### a. SVM по умолчанию
 
1. Обучите baseline-модель, только используя эти параметры: probability=True, kernel='linear', random_state=21.
2. Воспользуетесь K-fold кросс-валидацией со стратификацией, количество сплитов для оценки accuracy = 10.
3. Результат кода, в котором вы обучили модель и оценили ее качество, должен быть по формату точно такой же, как и в случае с логрегрессией.

In [24]:
%%time
cross_val(SVC(probability=True, kernel='linear', random_state=21), X_train, y_train, 10)

train  -  0.70486	|	valid  -  0.65926
train  -  0.69662	|	valid  -  0.75556
train  -  0.69415	|	valid  -  0.62222
train  -  0.70239	|	valid  -  0.65185
train  -  0.69085	|	valid  -  0.65185
train  -  0.68920	|	valid  -  0.64444
train  -  0.69250	|	valid  -  0.72593
train  -  0.70074	|	valid  -  0.62222
train  -  0.69605	|	valid  -  0.61940
train  -  0.71087	|	valid  -  0.63433
Average accuracy on crossval is 0.65871
Std is 0.04359
CPU times: total: 2.95 s
Wall time: 3.07 s


### b. Оптимизация параметров регуляризации
 В ячейках ниже попробуйте разные значения параметра C. Посмотрите, как меняется качество модели.

In [25]:
for C in (0.0001, 0.001, 0.01, 0.1, 1, 10, 100, 1000, 10000):
    print(f'C = {C}')
    cross_val(SVC(probability=True, kernel='linear', random_state=21, C=C), X_train, y_train, 10)
    print()

C = 0.0001
train  -  0.23495	|	valid  -  0.22963
train  -  0.23495	|	valid  -  0.22963
train  -  0.23495	|	valid  -  0.22963
train  -  0.23495	|	valid  -  0.22963
train  -  0.23413	|	valid  -  0.23704
train  -  0.23413	|	valid  -  0.23704
train  -  0.23413	|	valid  -  0.23704
train  -  0.23413	|	valid  -  0.23704
train  -  0.23394	|	valid  -  0.23881
train  -  0.23394	|	valid  -  0.23881
Average accuracy on crossval is 0.23443
Std is 0.00397
C = 0.001
train  -  0.23495	|	valid  -  0.22963
train  -  0.23495	|	valid  -  0.22963
train  -  0.23495	|	valid  -  0.22963
train  -  0.23495	|	valid  -  0.22963
train  -  0.23413	|	valid  -  0.23704
train  -  0.23413	|	valid  -  0.23704
train  -  0.23413	|	valid  -  0.23704
train  -  0.23413	|	valid  -  0.23704
train  -  0.23394	|	valid  -  0.23881
train  -  0.23394	|	valid  -  0.23881
Average accuracy on crossval is 0.23443
Std is 0.00397
C = 0.01
train  -  0.37923	|	valid  -  0.40000
train  -  0.37923	|	valid  -  0.40000
train  -  0.38417	|	vali

Best SVM: C=100  
Average accuracy on crossval is 0.75589
Std is 0.03550

### 4. Дерево
### a. Дерево по умолчанию
 
1. Обучите baseline-модель, только используя эти параметры: max_depth=10, random_state=21.
2. Воспользуетесь K-fold кросс-валидацией со стратификацией, количество сплитов для оценки accuracy = 10.
3. Результат кода, в котором вы обучили модель и оценили ее качество, должен быть по формату точно такой же, как и в случае с логрегрессией.

In [26]:
%%time
cross_val(DecisionTreeClassifier(max_depth=10, random_state=21), X_train, y_train, 10)

train  -  0.81039	|	valid  -  0.74074
train  -  0.77741	|	valid  -  0.74074
train  -  0.83347	|	valid  -  0.70370
train  -  0.79720	|	valid  -  0.76296
train  -  0.82440	|	valid  -  0.75556
train  -  0.80379	|	valid  -  0.68889
train  -  0.80709	|	valid  -  0.76296
train  -  0.80132	|	valid  -  0.65926
train  -  0.80807	|	valid  -  0.75373
train  -  0.80478	|	valid  -  0.68657
Average accuracy on crossval is 0.72551
Std is 0.03562
CPU times: total: 93.8 ms
Wall time: 81.3 ms


 ### b. Оптимизация параметров регуляризации
 
1. В ячейках ниже попробуйте разные значения параметра max_depth. 
2. В качестве бонуса попробуйте поиграться и с другими параметрами регуляризации. Посмотрите, как меняется качество модели.

In [27]:
for max_depth in range (12, 144, 12):
    print(f'max_depth = {max_depth}')
    print(cross_val(DecisionTreeClassifier(max_depth=max_depth, random_state=21), X_train, y_train, 10))
    print()

max_depth = 12
train  -  0.90272	|	valid  -  0.82963
train  -  0.85985	|	valid  -  0.83704
train  -  0.90932	|	valid  -  0.82222
train  -  0.89035	|	valid  -  0.85185
train  -  0.90932	|	valid  -  0.82963
train  -  0.86232	|	valid  -  0.71852
train  -  0.86562	|	valid  -  0.81481
train  -  0.88706	|	valid  -  0.75556
train  -  0.86820	|	valid  -  0.80597
train  -  0.87150	|	valid  -  0.77612
Average accuracy on crossval is 0.80413
Std is 0.03947
None

max_depth = 24
train  -  1.00000	|	valid  -  0.85926
train  -  0.99835	|	valid  -  0.92593
train  -  0.99753	|	valid  -  0.86667
train  -  0.99918	|	valid  -  0.91111
train  -  0.99835	|	valid  -  0.88148
train  -  0.99670	|	valid  -  0.86667
train  -  0.99588	|	valid  -  0.92593
train  -  0.99835	|	valid  -  0.88148
train  -  0.99918	|	valid  -  0.87313
train  -  0.99753	|	valid  -  0.88060
Average accuracy on crossval is 0.88722
Std is 0.02346
None

max_depth = 36
train  -  1.00000	|	valid  -  0.85926
train  -  1.00000	|	valid  -  0.918

In [28]:
for max_depth in range(12, 60, 12):
    for max_features in ('auto', 'sqrt', 'log2', None):
        print(f'max_depth = {max_depth}, max_features = {max_features}')
        print(cross_val(DecisionTreeClassifier(max_depth=max_depth, max_features=max_features, random_state=21), X_train, y_train, 10))
        print()

max_depth = 12, max_features = auto
train  -  0.75927	|	valid  -  0.70370
train  -  0.74031	|	valid  -  0.69630
train  -  0.75680	|	valid  -  0.69630
train  -  0.76917	|	valid  -  0.73333
train  -  0.73207	|	valid  -  0.64444
train  -  0.78813	|	valid  -  0.72593
train  -  0.79637	|	valid  -  0.76296
train  -  0.74938	|	valid  -  0.62963
train  -  0.75783	|	valid  -  0.70149
train  -  0.72323	|	valid  -  0.59701
Average accuracy on crossval is 0.68911
Std is 0.04822
None

max_depth = 12, max_features = sqrt
train  -  0.75927	|	valid  -  0.70370
train  -  0.74031	|	valid  -  0.69630
train  -  0.75680	|	valid  -  0.69630
train  -  0.76917	|	valid  -  0.73333
train  -  0.73207	|	valid  -  0.64444
train  -  0.78813	|	valid  -  0.72593
train  -  0.79637	|	valid  -  0.76296
train  -  0.74938	|	valid  -  0.62963
train  -  0.75783	|	valid  -  0.70149
train  -  0.72323	|	valid  -  0.59701
Average accuracy on crossval is 0.68911
Std is 0.04822
None

max_depth = 12, max_features = log2
train  -  

Best DecisionTreeClassifier: max_depth = 24  
Average accuracy on crossval is 0.88722  
Std is 0.02346

### 5. Случайный лес
### a. Случайный лес по умолчанию
 
1. Обучите baseline-модель, только используя эти параметры: n_estimators=50, max_depth=14, random_state=21.
2. Воспользуетесь K-fold кросс-валидацией со стратификацией, количество сплитов для оценки accuracy = 10.
3. Результат кода, в котором вы обучили модель и оценили ее качество, должен быть по формату точно такой же, как и в случае с логрегрессией.

In [29]:
%%time
cross_val(RandomForestClassifier(n_estimators=50, max_depth=14, random_state=21), X_train, y_train, 10)

train  -  0.96455	|	valid  -  0.88148
train  -  0.96208	|	valid  -  0.91852
train  -  0.96785	|	valid  -  0.86667
train  -  0.96455	|	valid  -  0.89630
train  -  0.96538	|	valid  -  0.91111
train  -  0.96538	|	valid  -  0.88148
train  -  0.97115	|	valid  -  0.91852
train  -  0.96867	|	valid  -  0.85185
train  -  0.97364	|	valid  -  0.88060
train  -  0.97941	|	valid  -  0.86567
Average accuracy on crossval is 0.88722
Std is 0.02204
CPU times: total: 812 ms
Wall time: 841 ms


### b. Оптимизация параметров регуляризации
 
1. В ячейках ниже попробуйте разные значения параметра max_depth и n_estimators. 
2. В качестве бонуса попробуйте поиграться и с другими параметрами регуляризации. Посмотрите, как меняется качество модели.

In [32]:
for max_depth in range(5, 26, 5):
    for n_estimators in range(10, 101, 10):
        print(f'max_depth = {max_depth}, n_estimators = {n_estimators}')
        print(cross_val(RandomForestClassifier(max_depth=max_depth, n_estimators=n_estimators, random_state=21), X_train, y_train, 10))
        print()

max_depth = 5, n_estimators = 10
train  -  0.60511	|	valid  -  0.54815
train  -  0.59110	|	valid  -  0.54815
train  -  0.57626	|	valid  -  0.53333
train  -  0.60676	|	valid  -  0.59259
train  -  0.58285	|	valid  -  0.55556
train  -  0.64138	|	valid  -  0.58519
train  -  0.61088	|	valid  -  0.57037
train  -  0.61253	|	valid  -  0.54074
train  -  0.63674	|	valid  -  0.57463
train  -  0.59061	|	valid  -  0.57463
Average accuracy on crossval is 0.56233
Std is 0.01887
None

max_depth = 5, n_estimators = 20
train  -  0.60264	|	valid  -  0.60741
train  -  0.55153	|	valid  -  0.49630
train  -  0.59357	|	valid  -  0.53333
train  -  0.59275	|	valid  -  0.57778
train  -  0.57873	|	valid  -  0.54074
train  -  0.60758	|	valid  -  0.56296
train  -  0.59687	|	valid  -  0.57037
train  -  0.60429	|	valid  -  0.56296
train  -  0.60297	|	valid  -  0.57463
train  -  0.58649	|	valid  -  0.55224
Average accuracy on crossval is 0.55787
Std is 0.02839
None

max_depth = 5, n_estimators = 30
train  -  0.60511	|

Best RandomForestClassifier: max_depth = 25, n_estimators = 90 
Average accuracy on crossval is 0.91840  
Std is 0.02120

### 6. Прогнозы
 
1. Выберите из всех вариантов лучшую модель и сделайте прогноз для тестовой выборки.
2. Посчитайте финальное accuracy.
3. Проанализируйте: для какого дня недели модель совершила наибольшее число ошибок (в % от общего количества наблюдений этого класса в вашем датасете).
4. Сохраните модель.

In [33]:
best = RandomForestClassifier(max_depth=25, n_estimators=90, random_state=21)
best.fit(X_train, y_train)

RandomForestClassifier(max_depth=25, n_estimators=90, random_state=21)

In [34]:
accuracy_score(y_test, best.predict(X_test))

0.9289940828402367

In [35]:
for day in range(7):
    print(day, sum((y_test == best.predict(X_test)) & (y_test == day)) / sum(y_test == day))

0 0.7407407407407407
1 0.9090909090909091
2 0.9333333333333333
3 0.975
4 0.8571428571428571
5 0.9259259259259259
6 0.9859154929577465


модель чаще всего ошибалась для понедельников, 74% правильных ответов

In [36]:
from joblib import dump, load

In [37]:
dump(best, 'best.joblib')

['best.joblib']

In [38]:
best1 = load('best.joblib')

In [39]:
accuracy_score(y_test, best1.predict(X_test))

0.9289940828402367